File: data-structures/linkedlist.py
"""
Comprehensive Linked List Implementations in Python

This module implements four linked list variants:
1. Singly Linked List
2. Doubly Linked List
3. Circular Linked List
4. Skip List (Probabilistic Data Structure)

Features:
- Basic operations: insert, delete, search, traverse
- Advanced operations: reverse, merge, split, cycle detection
- Iterator/generator support
- Memory efficient implementations
- Performance comparisons with arrays

Time Complexity Comparison:
                    Array   Linked List
Access by index:    O(1)    O(n)
Insert at head:     O(n)    O(1)
Insert at tail:     O(1)*   O(n) or O(1) with tail pointer
Insert at middle:   O(n)    O(n)
Delete at head:     O(n)    O(1)
Delete at tail:     O(1)*   O(n) or O(1) for doubly linked
Search:             O(n)    O(n)
* with dynamic arrays

Use Cases for Linked Lists:
- Frequent insertions/deletions at beginning
- Unknown or dynamic size
- Implementation of stacks, queues, graphs
- LRU caches
- Undo functionality
- Music playlists
- Browser history (doubly linked)

Cache Performance:
- Arrays have better cache locality (contiguous memory)
- Linked lists have poor spatial locality (scattered nodes)
- Arrays preferred for read-heavy workloads
- Linked lists preferred for write-heavy workloads at ends
"""

from typing import Generic, TypeVar, Optional, Iterator, List, Callable
import random

T = TypeVar('T')

# ============================================================================
# SINGLY LINKED LIST
# ============================================================================

class SinglyNode(Generic[T]):
    """Node for singly linked list."""

    def __init__(self, data: T):
        self.data = data
        self.next: Optional['SinglyNode[T]'] = None


class SinglyLinkedList(Generic[T]):
    """
    Singly Linked List - each node points to next node only.

    Advantages:
    - Simple implementation
    - Less memory per node (one pointer)
    - O(1) insertion at head

    Disadvantages:
    - Can only traverse forward
    - O(n) to find previous node
    - O(n) deletion at tail
    """

    def __init__(self):
        self.head: Optional[SinglyNode[T]] = None
        self.size = 0

    def insert_at_head(self, data: T) -> None:
        """Insert at beginning. O(1)"""
        new_node = SinglyNode(data)
        new_node.next = self.head
        self.head = new_node
        self.size += 1

    def insert_at_tail(self, data: T) -> None:
        """Insert at end. O(n)"""
        new_node = SinglyNode(data)

        if not self.head:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node

        self.size += 1

    def insert_at_position(self, data: T, position: int) -> None:
        """Insert at specific position. O(n)"""
        if position < 0 or position > self.size:
            raise IndexError("Position out of bounds")

        if position == 0:
            self.insert_at_head(data)
            return

        new_node = SinglyNode(data)
        current = self.head

        for _ in range(position - 1):
            current = current.next

        new_node.next = current.next
        current.next = new_node
        self.size += 1

    def delete_at_head(self) -> Optional[T]:
        """Delete from beginning. O(1)"""
        if not self.head:
            return None

        data = self.head.data
        self.head = self.head.next
        self.size -= 1
        return data

    def delete_at_tail(self) -> Optional[T]:
        """Delete from end. O(n)"""
        if not self.head:
            return None

        if not self.head.next:
            data = self.head.data
            self.head = None
            self.size -= 1
            return data

        current = self.head
        while current.next.next:
            current = current.next

        data = current.next.data
        current.next = None
        self.size -= 1
        return data

    def delete_value(self, value: T) -> bool:
        """Delete first occurrence of value. O(n)"""
        if not self.head:
            return False

        if self.head.data == value:
            self.head = self.head.next
            self.size -= 1
            return True

        current = self.head
        while current.next:
            if current.next.data == value:
                current.next = current.next.next
                self.size -= 1
                return True
            current = current.next

        return False

    def search(self, value: T) -> Optional[SinglyNode[T]]:
        """Search for value. O(n)"""
        current = self.head
        while current:
            if current.data == value:
                return current
            current = current.next
        return None

    def reverse(self) -> None:
        """Reverse the list in-place. O(n)"""
        prev = None
        current = self.head

        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node

        self.head = prev

    def get_middle(self) -> Optional[T]:
        """Get middle element using slow/fast pointers. O(n)"""
        if not self.head:
            return None

        slow = fast = self.head

        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next

        return slow.data

    def detect_cycle(self) -> bool:
        """Detect cycle using Floyd's algorithm. O(n)"""
        if not self.head:
            return False

        slow = fast = self.head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True

        return False

    def merge_sorted(self, other: 'SinglyLinkedList[T]') -> 'SinglyLinkedList[T]':
        """Merge two sorted lists. O(n + m)"""
        result = SinglyLinkedList[T]()

        current1 = self.head
        current2 = other.head

        # Find the head of merged list
        if not current1:
            result.head = current2
            result.size = other.size
            return result
        if not current2:
            result.head = current1
            result.size = self.size
            return result

        if current1.data <= current2.data:
            result.head = current1
            current1 = current1.next
        else:
            result.head = current2
            current2 = current2.next

        current = result.head

        while current1 and current2:
            if current1.data <= current2.data:
                current.next = current1
                current1 = current1.next
            else:
                current.next = current2
                current2 = current2.next
            current = current.next

        current.next = current1 if current1 else current2
        result.size = self.size + other.size

        return result

    def __iter__(self) -> Iterator[T]:
        """Iterator support."""
        current = self.head
        while current:
            yield current.data
            current = current.next

    def __len__(self) -> int:
        return self.size

    def __str__(self) -> str:
        return ' -> '.join(str(data) for data in self) + ' -> None'


# ============================================================================
# DOUBLY LINKED LIST
# ============================================================================

class DoublyNode(Generic[T]):
    """Node for doubly linked list."""

    def __init__(self, data: T):
        self.data = data
        self.next: Optional['DoublyNode[T]'] = None
        self.prev: Optional['DoublyNode[T]'] = None


class DoublyLinkedList(Generic[T]):
    """
    Doubly Linked List - each node points to both next and previous.

    Advantages:
    - Bidirectional traversal
    - O(1) deletion at tail (with tail pointer)
    - Easier deletion of specific node
    - Better for implementing deque

    Disadvantages:
    - More memory per node (two pointers)
    - Slightly more complex operations
    """

    def __init__(self):
        self.head: Optional[DoublyNode[T]] = None
        self.tail: Optional[DoublyNode[T]] = None
        self.size = 0

    def insert_at_head(self, data: T) -> None:
        """Insert at beginning. O(1)"""
        new_node = DoublyNode(data)

        if not self.head:
            self.head = self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node

        self.size += 1

    def insert_at_tail(self, data: T) -> None:
        """Insert at end. O(1)"""
        new_node = DoublyNode(data)

        if not self.tail:
            self.head = self.tail = new_node
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node

        self.size += 1

    def delete_at_head(self) -> Optional[T]:
        """Delete from beginning. O(1)"""
        if not self.head:
            return None

        data = self.head.data

        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.head = self.head.next
            self.head.prev = None

        self.size -= 1
        return data

    def delete_at_tail(self) -> Optional[T]:
        """Delete from end. O(1)"""
        if not self.tail:
            return None

        data = self.tail.data

        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.tail = self.tail.prev
            self.tail.next = None

        self.size -= 1
        return data

    def delete_node(self, node: DoublyNode[T]) -> None:
        """Delete specific node. O(1) with node reference"""
        if node.prev:
            node.prev.next = node.next
        else:
            self.head = node.next

        if node.next:
            node.next.prev = node.prev
        else:
            self.tail = node.prev

        self.size -= 1

    def reverse(self) -> None:
        """Reverse the list in-place. O(n)"""
        current = self.head
        self.head, self.tail = self.tail, self.head

        while current:
            current.prev, current.next = current.next, current.prev
            current = current.prev  # Move to next (which is now prev)

    def __iter__(self) -> Iterator[T]:
        """Forward iterator."""
        current = self.head
        while current:
            yield current.data
            current = current.next

    def reverse_iter(self) -> Iterator[T]:
        """Backward iterator."""
        current = self.tail
        while current:
            yield current.data
            current = current.prev

    def __len__(self) -> int:
        return self.size

    def __str__(self) -> str:
        return ' <-> '.join(str(data) for data in self) + ' <-> None'


# ============================================================================
# CIRCULAR LINKED LIST
# ============================================================================

class CircularLinkedList(Generic[T]):
    """
    Circular Linked List - last node points back to first.

    Advantages:
    - Can traverse entire list from any node
    - Useful for round-robin scheduling
    - No null pointers
    - Natural for cyclic data

    Disadvantages:
    - Need special care to avoid infinite loops
    - Slightly more complex operations

    Use Cases:
    - Round-robin CPU scheduling
    - Multiplayer game turns
    - Music playlist on repeat
    - Circular buffers
    """

    def __init__(self):
        self.head: Optional[SinglyNode[T]] = None
        self.size = 0

    def insert_at_head(self, data: T) -> None:
        """Insert at beginning. O(n) to find last node"""
        new_node = SinglyNode(data)

        if not self.head:
            new_node.next = new_node  # Points to itself
            self.head = new_node
        else:
            # Find last node
            current = self.head
            while current.next != self.head:
                current = current.next

            new_node.next = self.head
            current.next = new_node
            self.head = new_node

        self.size += 1

    def insert_at_tail(self, data: T) -> None:
        """Insert at end. O(n)"""
        new_node = SinglyNode(data)

        if not self.head:
            new_node.next = new_node
            self.head = new_node
        else:
            current = self.head
            while current.next != self.head:
                current = current.next

            current.next = new_node
            new_node.next = self.head

        self.size += 1

    def delete_at_head(self) -> Optional[T]:
        """Delete from beginning. O(n)"""
        if not self.head:
            return None

        data = self.head.data

        if self.head.next == self.head:
            self.head = None
        else:
            # Find last node
            current = self.head
            while current.next != self.head:
                current = current.next

            current.next = self.head.next
            self.head = self.head.next

        self.size -= 1
        return data

    def traverse(self, callback: Callable[[T], None]) -> None:
        """Traverse and apply callback. O(n)"""
        if not self.head:
            return

        current = self.head
        while True:
            callback(current.data)
            current = current.next
            if current == self.head:
                break

    def __iter__(self) -> Iterator[T]:
        """Iterator support."""
        if not self.head:
            return

        current = self.head
        while True:
            yield current.data
            current = current.next
            if current == self.head:
                break

    def __len__(self) -> int:
        return self.size

    def __str__(self) -> str:
        if not self.head:
            return "Empty"
        return ' -> '.join(str(data) for data in self) + ' -> (head)'


# ============================================================================
# SKIP LIST
# ============================================================================

class SkipNode(Generic[T]):
    """Node for skip list with multiple forward pointers."""

    def __init__(self, data: T, level: int):
        self.data = data
        self.forward: List[Optional['SkipNode[T]']] = [None] * (level + 1)


class SkipList(Generic[T]):
    """
    Skip List - probabilistic data structure for fast search.

    Complexity:
    - Search: O(log n) average
    - Insert: O(log n) average
    - Delete: O(log n) average
    - Space: O(n log n) expected

    Advantages:
    - Faster than regular linked list for search
    - Simpler than balanced trees
    - Probabilistic balancing (no rotations)
    - Good for concurrent access

    Disadvantages:
    - More memory overhead
    - Probabilistic (worst case still O(n))
    - Cache performance worse than B-trees

    Use Cases:
    - In-memory databases (Redis uses skip lists)
    - Concurrent data structures
    - Alternative to balanced trees
    - Level-based indexing
    """

    MAX_LEVEL = 16
    P = 0.5  # Probability factor

    def __init__(self):
        self.level = 0
        self.header = SkipNode(None, self.MAX_LEVEL)
        self.size = 0

    def _random_level(self) -> int:
        """Generate random level using probability P."""
        level = 0
        while random.random() < self.P and level < self.MAX_LEVEL:
            level += 1
        return level

    def insert(self, data: T) -> None:
        """Insert element. O(log n) average"""
        update = [None] * (self.MAX_LEVEL + 1)
        current = self.header

        # Find position to insert
        for i in range(self.level, -1, -1):
            while current.forward[i] and current.forward[i].data < data:
                current = current.forward[i]
            update[i] = current

        # Generate random level
        new_level = self._random_level()

        if new_level > self.level:
            for i in range(self.level + 1, new_level + 1):
                update[i] = self.header
            self.level = new_level

        # Create new node
        new_node = SkipNode(data, new_level)

        # Update forward pointers
        for i in range(new_level + 1):
            new_node.forward[i] = update[i].forward[i]
            update[i].forward[i] = new_node

        self.size += 1

    def search(self, data: T) -> bool:
        """Search for element. O(log n) average"""
        current = self.header

        for i in range(self.level, -1, -1):
            while current.forward[i] and current.forward[i].data < data:
                current = current.forward[i]

        current = current.forward[0]
        return current is not None and current.data == data

    def delete(self, data: T) -> bool:
        """Delete element. O(log n) average"""
        update = [None] * (self.MAX_LEVEL + 1)
        current = self.header

        # Find node to delete
        for i in range(self.level, -1, -1):
            while current.forward[i] and current.forward[i].data < data:
                current = current.forward[i]
            update[i] = current

        current = current.forward[0]

        if current is None or current.data != data:
            return False

        # Remove node
        for i in range(self.level + 1):
            if update[i].forward[i] != current:
                break
            update[i].forward[i] = current.forward[i]

        # Update level
        while self.level > 0 and self.header.forward[self.level] is None:
            self.level -= 1

        self.size -= 1
        return True

    def __iter__(self) -> Iterator[T]:
        """Iterator at lowest level."""
        current = self.header.forward[0]
        while current:
            yield current.data
            current = current.forward[0]

    def __len__(self) -> int:
        return self.size

    def __str__(self) -> str:
        return f"SkipList({list(self)})"


# ============================================================================
# DEMONSTRATION AND BENCHMARKING
# ============================================================================

def demonstrate_linked_lists():
    """Demonstrate all linked list variants."""

    print("=" * 80)
    print("COMPREHENSIVE LINKED LIST DEMONSTRATIONS")
    print("=" * 80)

    # Singly Linked List
    print("\n1. SINGLY LINKED LIST")
    print("-" * 80)
    sll = SinglyLinkedList[int]()

    print("Inserting: 1, 2, 3 at head")
    sll.insert_at_head(3)
    sll.insert_at_head(2)
    sll.insert_at_head(1)
    print(f"List: {sll}")

    print("\nInserting: 4, 5 at tail")
    sll.insert_at_tail(4)
    sll.insert_at_tail(5)
    print(f"List: {sll}")

    print(f"\nMiddle element: {sll.get_middle()}")

    print("\nReversing list...")
    sll.reverse()
    print(f"List: {sll}")

    # Doubly Linked List
    print("\n2. DOUBLY LINKED LIST")
    print("-" * 80)
    dll = DoublyLinkedList[str]()

    print("Inserting: A, B, C at head")
    dll.insert_at_head("C")
    dll.insert_at_head("B")
    dll.insert_at_head("A")
    print(f"List: {dll}")

    print("\nInserting: D, E at tail")
    dll.insert_at_tail("D")
    dll.insert_at_tail("E")
    print(f"List: {dll}")

    print("\nForward iteration:", list(dll))
    print("Backward iteration:", list(dll.reverse_iter()))

    print("\nDeleting head and tail...")
    dll.delete_at_head()
    dll.delete_at_tail()
    print(f"List: {dll}")

    # Circular Linked List
    print("\n3. CIRCULAR LINKED LIST")
    print("-" * 80)
    cll = CircularLinkedList[int]()

    print("Inserting: 1, 2, 3, 4, 5")
    for i in range(1, 6):
        cll.insert_at_tail(i)
    print(f"List: {cll}")

    print("\nTraversing (will loop back):")
    count = 0
    for item in cll:
        print(f"  {item}", end=" ")
        count += 1
        if count >= 8:  # Show it loops
            print("...")
            break
    print()

    # Skip List
    print("\n4. SKIP LIST")
    print("-" * 80)
    sl = SkipList[int]()

    print("Inserting: 3, 7, 1, 9, 5, 2, 8, 4, 6")
    for val in [3, 7, 1, 9, 5, 2, 8, 4, 6]:
        sl.insert(val)

    print(f"Skip list (sorted): {list(sl)}")

    print("\nSearching for 5:", sl.search(5))
    print("Searching for 10:", sl.search(10))

    print("\nDeleting 5...")
    sl.delete(5)
    print(f"Skip list: {list(sl)}")

    # Performance comparison
    print("\n5. PERFORMANCE COMPARISON: LINKED LIST VS ARRAY")
    print("-" * 80)
    print("""
    Operation           Array       Singly LL   Doubly LL   Skip List
    ----------------------------------------------------------------
    Access by index     O(1)        O(n)        O(n)        O(log n)
    Insert at head      O(n)        O(1)        O(1)        O(log n)
    Insert at tail      O(1)*       O(n)        O(1)        O(log n)
    Delete at head      O(n)        O(1)        O(1)        O(log n)
    Delete at tail      O(1)*       O(n)        O(1)        O(log n)
    Search              O(n)        O(n)        O(n)        O(log n)
    Memory overhead     Low         Medium      High        Very High
    Cache locality      Excellent   Poor        Poor        Poor

    * Amortized for dynamic arrays

    WHEN TO USE LINKED LISTS:
    ✓ Frequent insertions/deletions at beginning or end
    ✓ Size unknown or highly variable
    ✓ Don't need random access
    ✓ Implementing stacks, queues, or graphs
    ✓ Memory fragmentation acceptable

    WHEN TO USE ARRAYS:
    ✓ Random access needed
    ✓ Read-heavy workloads
    ✓ Size known or slowly growing
    ✓ Cache performance critical
    ✓ Memory locality important
    """)

    print("\n6. USE CASE EXAMPLES")
    print("-" * 80)
    print("""
    Singly Linked List:
    - Stack implementation
    - Simple queues
    - Hash table chaining
    - Undo functionality

    Doubly Linked List:
    - Browser history (back/forward)
    - LRU cache
    - Deque implementation
    - Music player (prev/next)
    - Text editor (cursor movement)

    Circular Linked List:
    - Round-robin scheduling
    - Multiplayer games (turn order)
    - Circular buffers
    - Music playlist on repeat

    Skip List:
    - In-memory databases (Redis)
    - Concurrent data structures
    - Sorted collections
    - Alternative to balanced trees
    """)

    print("=" * 80)
    print("✨ All demonstrations complete!")
    print("=" * 80)


if __name__ == "__main__":
    demonstrate_linked_lists()

File: data-structures/linkedlist.rb
# Comprehensive Linked List Implementations in Ruby
#
# Features:
# - Dynamic typing with duck typing
# - Block/Proc support for callbacks
# - All 4 variants: Singly, Doubly, Circular, Skip List
# - Enumerable module inclusion
#
# Usage:
#   ruby linkedlist.rb

# ============================================================================
# SINGLY LINKED LIST
# ============================================================================

class SinglyNode
  attr_accessor :data, :next

  def initialize(data)
    @data = data
    @next = nil
  end
end

class SinglyLinkedList
  include Enumerable

  attr_reader :size

  def initialize
    @head = nil
    @size = 0
  end

  def insert_at_head(data)
    new_node = SinglyNode.new(data)
    new_node.next = @head
    @head = new_node
    @size += 1
  end

  def insert_at_tail(data)
    new_node = SinglyNode.new(data)

    if @head.nil?
      @head = new_node
    else
      current = @head
      current = current.next while current.next
      current.next = new_node
    end

    @size += 1
  end

  def delete_at_head
    return nil if @head.nil?

    data = @head.data
    @head = @head.next
    @size -= 1
    data
  end

  def search(value = nil, &block)
    current = @head

    while current
      if block_given?
        return current.data if yield(current.data)
      else
        return current.data if current.data == value
      end
      current = current.next
    end

    nil
  end

  def reverse!
    prev = nil
    current = @head

    while current
      next_node = current.next
      current.next = prev
      prev = current
      current = next_node
    end

    @head = prev
    self
  end

  def get_middle
    return nil if @head.nil?

    slow = @head
    fast = @head

    while fast.next && fast.next.next
      slow = slow.next
      fast = fast.next.next
    end

    slow.data
  end

  def detect_cycle?
    return false if @head.nil?

    slow = @head
    fast = @head

    while fast && fast.next
      slow = slow.next
      fast = fast.next.next
      return true if slow == fast
    end

    false
  end

  # Enumerable support
  def each
    return enum_for(:each) unless block_given?

    current = @head
    while current
      yield current.data
      current = current.next
    end
  end

  def empty?
    @head.nil?
  end

  def to_s
    map(&:to_s).join(' -> ') + ' -> nil'
  end

  alias_method :inspect, :to_s
end

# ============================================================================
# DOUBLY LINKED LIST
# ============================================================================

class DoublyNode
  attr_accessor :data, :next, :prev

  def initialize(data)
    @data = data
    @next = nil
    @prev = nil
  end
end

class DoublyLinkedList
  include Enumerable

  attr_reader :size

  def initialize
    @head = nil
    @tail = nil
    @size = 0
  end

  def insert_at_head(data)
    new_node = DoublyNode.new(data)

    if @head.nil?
      @head = @tail = new_node
    else
      new_node.next = @head
      @head.prev = new_node
      @head = new_node
    end

    @size += 1
  end

  def insert_at_tail(data)
    new_node = DoublyNode.new(data)

    if @tail.nil?
      @head = @tail = new_node
    else
      new_node.prev = @tail
      @tail.next = new_node
      @tail = new_node
    end

    @size += 1
  end

  def delete_at_head
    return nil if @head.nil?

    data = @head.data

    if @head == @tail
      @head = @tail = nil
    else
      @head = @head.next
      @head.prev = nil
    end

    @size -= 1
    data
  end

  def delete_at_tail
    return nil if @tail.nil?

    data = @tail.data

    if @head == @tail
      @head = @tail = nil
    else
      @tail = @tail.prev
      @tail.next = nil
    end

    @size -= 1
    data
  end

  def reverse!
    current = @head
    @head, @tail = @tail, @head

    while current
      current.prev, current.next = current.next, current.prev
      current = current.prev
    end

    self
  end

  # Forward iteration (default)
  def each
    return enum_for(:each) unless block_given?

    current = @head
    while current
      yield current.data
      current = current.next
    end
  end

  # Backward iteration
  def reverse_each
    return enum_for(:reverse_each) unless block_given?

    current = @tail
    while current
      yield current.data
      current = current.prev
    end
  end

  def empty?
    @head.nil?
  end

  def to_s
    map(&:to_s).join(' <-> ') + ' <-> nil'
  end

  alias_method :inspect, :to_s
end

# ============================================================================
# CIRCULAR LINKED LIST
# ============================================================================

class CircularLinkedList
  attr_reader :size

  def initialize
    @head = nil
    @size = 0
  end

  def insert_at_tail(data)
    new_node = SinglyNode.new(data)

    if @head.nil?
      new_node.next = new_node
      @head = new_node
    else
      current = @head
      current = current.next while current.next != @head

      current.next = new_node
      new_node.next = @head
    end

    @size += 1
  end

  def delete_at_head
    return nil if @head.nil?

    data = @head.data

    if @head.next == @head
      @head = nil
    else
      current = @head
      current = current.next while current.next != @head

      current.next = @head.next
      @head = @head.next
    end

    @size -= 1
    data
  end

  def each
    return enum_for(:each) unless block_given?
    return if @head.nil?

    current = @head
    loop do
      yield current.data
      current = current.next
      break if current == @head
    end
  end

  def empty?
    @head.nil?
  end

  def to_s
    return 'Empty' if @head.nil?

    elements = []
    current = @head

    loop do
      elements << current.data.to_s
      current = current.next
      break if current == @head
    end

    elements.join(' -> ') + ' -> (head)'
  end

  alias_method :inspect, :to_s
end

# ============================================================================
# SKIP LIST
# ============================================================================

class SkipNode
  attr_accessor :data, :forward

  def initialize(data, level)
    @data = data
    @forward = Array.new(level + 1)
  end
end

class SkipList
  MAX_LEVEL = 16
  P = 0.5

  attr_reader :size

  def initialize
    @header = SkipNode.new(nil, MAX_LEVEL)
    @level = 0
    @size = 0
  end

  def random_level
    lvl = 0
    lvl += 1 while rand < P && lvl < MAX_LEVEL
    lvl
  end

  def insert(data)
    update = Array.new(MAX_LEVEL + 1)
    current = @header

    @level.downto(0) do |i|
      while current.forward[i] && current.forward[i].data < data
        current = current.forward[i]
      end
      update[i] = current
    end

    new_level = random_level

    if new_level > @level
      ((@level + 1)..new_level).each do |i|
        update[i] = @header
      end
      @level = new_level
    end

    new_node = SkipNode.new(data, new_level)

    (0..new_level).each do |i|
      new_node.forward[i] = update[i].forward[i]
      update[i].forward[i] = new_node
    end

    @size += 1
  end

  def search(data)
    current = @header

    @level.downto(0) do |i|
      while current.forward[i] && current.forward[i].data < data
        current = current.forward[i]
      end
    end

    current = current.forward[0]
    current && current.data == data
  end

  def delete(data)
    update = Array.new(MAX_LEVEL + 1)
    current = @header

    @level.downto(0) do |i|
      while current.forward[i] && current.forward[i].data < data
        current = current.forward[i]
      end
      update[i] = current
    end

    current = current.forward[0]

    return false unless current && current.data == data

    (0..@level).each do |i|
      break if update[i].forward[i] != current
      update[i].forward[i] = current.forward[i]
    end

    @level -= 1 while @level > 0 && @header.forward[@level].nil?

    @size -= 1
    true
  end

  def to_a
    result = []
    current = @header.forward[0]

    while current
      result << current.data
      current = current.forward[0]
    end

    result
  end

  def to_s
    "SkipList#{to_a}"
  end

  alias_method :inspect, :to_s
end

# ============================================================================
# DEMONSTRATION
# ============================================================================

def demonstrate
  puts '=' * 80
  puts 'COMPREHENSIVE LINKED LIST DEMONSTRATIONS IN RUBY'
  puts '=' * 80

  # Singly Linked List
  puts "\n1. SINGLY LINKED LIST"
  puts '-' * 80
  sll = SinglyLinkedList.new

  puts 'Inserting: 1, 2, 3 at head'
  sll.insert_at_head(3)
  sll.insert_at_head(2)
  sll.insert_at_head(1)
  puts "List: #{sll}"

  puts "\nInserting: 4, 5 at tail"
  sll.insert_at_tail(4)
  sll.insert_at_tail(5)
  puts "List: #{sll}"

  puts "\nMiddle element: #{sll.get_middle}"

  puts "\nReversing list..."
  sll.reverse!
  puts "List: #{sll}"

  puts "\nUsing blocks - find first even number:"
  even = sll.search { |x| x.even? }
  puts "Found: #{even}"

  # Doubly Linked List
  puts "\n2. DOUBLY LINKED LIST"
  puts '-' * 80
  dll = DoublyLinkedList.new

  puts 'Inserting: A, B, C at head'
  dll.insert_at_head('C')
  dll.insert_at_head('B')
  dll.insert_at_head('A')
  puts "List: #{dll}"

  puts "\nInserting: D, E at tail"
  dll.insert_at_tail('D')
  dll.insert_at_tail('E')
  puts "List: #{dll}"

  puts "\nForward iteration: #{dll.to_a}"
  puts "Backward iteration: #{dll.reverse_each.to_a}"

  # Circular Linked List
  puts "\n3. CIRCULAR LINKED LIST"
  puts '-' * 80
  cll = CircularLinkedList.new

  puts 'Inserting: 1, 2, 3, 4, 5'
  (1..5).each { |i| cll.insert_at_tail(i) }
  puts "List: #{cll}"

  # Skip List
  puts "\n4. SKIP LIST"
  puts '-' * 80
  sl = SkipList.new

  puts 'Inserting: 3, 7, 1, 9, 5, 2, 8, 4, 6'
  [3, 7, 1, 9, 5, 2, 8, 4, 6].each { |val| sl.insert(val) }

  puts "Skip list (sorted): #{sl}"

  puts "\nSearching for 5: #{sl.search(5)}"
  puts "Searching for 10: #{sl.search(10)}"

  puts "\nDeleting 5..."
  sl.delete(5)
  puts "Skip list: #{sl}"

  puts "\n#{'=' * 80}"
  puts '✨ All demonstrations complete!'
  puts '=' * 80
end

demonstrate if __FILE__ == $PROGRAM_NAME
